src/pipelines/step/composite/mod.rs

```diff
@@ -0,0 +1,148 @@
+/// It takes the list of steps, initial payload and context
+trait CompositeStepMode<P: Platform>: Send + Sync {
+	fn steps(
+		&self,
+		steps: &[Arc<StepInstance<P>>],
+		payload: Checkpoint<P>,
+		ctx: StepContext<P>,
+	) -> impl Future<Output = ControlFlow<P>> + Send;
+}
+
+/// `CompositeStep` will execute all before/after/setup functions of each step
+/// in order The step implementation is delegated to the composite mode
+impl<P, M> Step<P> for CompositeStep<P, M>
+where
+	P: Platform,
+	M: CompositeStepMode<P> + Send + 'static,
+{
+	async fn step(
+		self: Arc<Self>,
+		payload: Checkpoint<P>,
+		ctx: StepContext<P>,
+	) -> ControlFlow<P> {
+		self.mode.steps(&self.steps, payload, ctx).await
+	}
+
+	async fn before_job(
+		self: Arc<Self>,
+		ctx: StepContext<P>,
+	) -> Result<(), PayloadBuilderError> {
+		for step in &self.steps {
+			step.before_job(ctx.clone()).await?;
+		}
+		Ok(())
+	}
+
+	async fn after_job(
+		self: Arc<Self>,
+		ctx: StepContext<P>,
+		result: Arc<Result<BuiltPayload<P>, PayloadBuilderError>>,
+	) -> Result<(), PayloadBuilderError> {
+		for step in &self.steps {
+			step.after_job(ctx.clone(), result.clone()).await?;
+		}
+		Ok(())
+	}
+
+	async fn setup(
+		&mut self,
+		init: InitContext<P>,
+	) -> Result<(), PayloadBuilderError> {
+		for step in &self.steps {
+			step.setup(init.clone()).await?;
+		}
+		Ok(())
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use {super::*, crate::test_utils::*};
+
+	// Composite step mode that tries to execute all steps and ignore all failures
+	// and breaks Will return the latest successful result
+	struct TryAllMode;
+	impl<P: Platform> CompositeStepMode<P> for TryAllMode {
+		async fn steps(
+			&self,
+			steps: &[Arc<StepInstance<P>>],
+			payload: Checkpoint<P>,
+			ctx: StepContext<P>,
+		) -> ControlFlow<P> {
+			let mut current = payload.clone();
+			for step in steps {
+				if let ControlFlow::Ok(res) =
+					step.step(current.clone(), ctx.clone()).await
+				{
+					current = res;
+				}
+			}
+			ControlFlow::Ok(current)
+		}
+	}
+
+	#[rblib_test(Ethereum)]
+	async fn composite_empty_steps<P: TestablePlatform>() -> eyre::Result<()> {
+		let composite = CompositeStep::<P, _>::new(TryAllMode);
+		let result = OneStep::<P>::new(composite).run().await?;
+		assert!(matches!(result, ControlFlow::Ok(_)));
+		Ok(())
+	}
+
+	#[rblib_test(Ethereum)]
+	async fn composite_single_step_ok<P: TestablePlatform>() -> eyre::Result<()> {
+		let mut composite = CompositeStep::<P, _>::new(TryAllMode);
+		composite.append_step(AlwaysOkStep);
+
+		let result = OneStep::<P>::new(composite).run().await?;
+		assert!(matches!(result, ControlFlow::Ok(_)));
+		Ok(())
+	}
+
+	#[rblib_test(Ethereum)]
+	async fn composite_multiple_steps_ok<P: TestablePlatform>() -> eyre::Result<()>
+	{
+		let mut composite = CompositeStep::<P, _>::new(TryAllMode);
+		composite.append_step(AlwaysOkStep);
+		composite.append_step(AlwaysOkStep);
+		composite.append_step(AlwaysOkStep);
```

Comment:
I'd suggest that we tweak the DevEx here a little bit and enable the following syntax:

```rust
let pipeline = Pipeline::default()
  .with_prologue(All((Step1, Step2, Step3))
  .step(Step4)
  .with_epilogue(Any((Step5, Step6)));
```

This is how this syntax is implemented for constructing pipelines:

https://github.com/flashbots/rblib/blob/d52ed99c348865b587a0f9b62570b872af883a4e/src/pipelines/mod.rs#L226-L269


Not a blocker, but a very nice to have to preserve consistency with the overall syntax.